mlos_bench/mlos_bench/storage/base_storage.py

```diff
@@ -0,0 +1,231 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Base interface for saving and restoring the benchmark data.
+"""
+
+import logging
+from abc import ABCMeta, abstractmethod
+from typing import Optional, Union, List, Tuple, Dict, Any
+
+from mlos_bench.environment import Status
+from mlos_bench.service import Service
+from mlos_bench.tunables import TunableGroups
+from mlos_bench.util import get_git_info
+
+_LOG = logging.getLogger(__name__)
+
+# pylint: disable=too-few-public-methods,too-many-arguments
+
+
+class Storage(metaclass=ABCMeta):
+    """
+    An abstract interface between the benchmarking framework
+    and storage systems (e.g., SQLite or MLFLow).
+    """
+
+    def __init__(self, tunables: TunableGroups, service: Optional[Service], config: dict):
+        """
+        Create a new storage object.
+
+        Parameters
+        ----------
+        tunables : TunableGroups
+            Tunable parameters of the environment. We need them to validate the
+            configurations of merged-in experiments and restored/pending trials.
+        config : dict
+            Free-format key/value pairs of configuration parameters.
+        """
+        _LOG.debug("Storage config: %s", config)
+        self._tunables = tunables.copy()
+        self._service = service
+        self._config = config.copy()
+
+    @abstractmethod
+    def experiment(self, exp_id: str, trial_id: int, description: str, opt_target: str):
+        """
+        Create a new experiment in the storage.
+
+        Parameters
+        ----------
+        exp_id : str
+            Unique identifier of the experiment.
+        trial_id : int
+            Starting number of the trial.
+        description : str
+            Human-readable description of the experiment.
+        opt_target : str
+            Name of metric we're optimizing for.
```

Comment:
Also need the direction (min or max).

Reply:
we don't, actually. All we need here is a metric name to retrieve from storage; we don't care about the direction as long as we took the right measurements in the earlier trials.

Later we will pass the right metadata for multi-target optimization along with different directions for each metric, etc., but for now this is the absolute minimum we need to save and restore the trial data. We'll add more stuff in subsequent PRs

Reply:
Leaving multi-objective aside for now, since I agree that's future work, we do however already support minimization vs. maximizing the given objective target.
The framework needs to know that when restarting an experiment or merging them, correct?
Or is the assumption that we'll be able to pull that from the local config files still?
In that case, do we need to store the opt_target in the DB at all?

Reply:
I think we've resolved that during our conversation today. We don't store the `opt_target` in the DB now, and this parameter is just for sanity check to make sure we retrieve the right metric when we load the data from previous trials. I will add a comment explaining it in the code.